plugin/src/main/java/org/opensearch/sql/plugin/catalog/CatalogServiceImpl.java

```diff
@@ -0,0 +1,180 @@
+    }};
+
+  private Map<String, StorageEngine> storageEngineMap = new HashMap<>();
+
+  private static final CatalogServiceImpl INSTANCE = new CatalogServiceImpl();
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  public static CatalogServiceImpl getInstance() {
+    return INSTANCE;
+  }
+
+  private CatalogServiceImpl() {
+  }
+
+  /**
+   * This function reads settings and loads connectors to the data stores.
+   * This will be invoked during start up and also when settings are updated.
+   *
+   * @param settings settings.
+   */
+  public void loadConnectors(Settings settings) {
+    doPrivileged(() -> {
+      Boolean isPPLEnabled = (Boolean) OpenSearchSettings.PPL_ENABLED_SETTING.get(settings);
+      Boolean isFederationEnabled = CatalogSettings.FEDERATION_ENABLED.get(settings);
+      if (isPPLEnabled && isFederationEnabled) {
+        InputStream inputStream = CatalogSettings.CATALOG_CONFIG.get(settings);
+        if (inputStream != null) {
+          ObjectMapper objectMapper = new ObjectMapper();
+          objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+          try {
+            ArrayNode catalogs = objectMapper.readValue(inputStream, ArrayNode.class);
+            validateCatalogs(catalogs);
+            constructConnectors(catalogs);
+          } catch (IOException e) {
+            LOG.error("Catalog Configuration File uploaded is malformed. Verify and re-upload.");
+            throw new IllegalArgumentException(
+                "Malformed Catalog Configuration Json" + e.getMessage());
+          }
+        }
+      } else {
+        LOG.info("PPL federation is not enabled.");
+      }
+      return null;
+    });
+  }
+
+  @Override
+  public Optional<StorageEngine> getStorageEngine(String catalog) {
+    return Optional.ofNullable(storageEngineMap.get(catalog));
+  }
+
+  @Override
+  public Set<String> getCatalogs() {
+    return storageEngineMap.keySet();
+  }
+
+  private <T> T doPrivileged(PrivilegedExceptionAction<T> action) {
+    try {
+      return SecurityAccess.doPrivileged(action);
+    } catch (IOException e) {
+      throw new IllegalStateException("Failed to perform privileged action", e);
+    }
+  }
+
+  private StorageEngine createStorageEngine(JsonNode catalog) throws URISyntaxException {
+    StorageEngine storageEngine;
+    String connector = catalog.get(CatalogConstants.CONNECTOR).asText();
+    switch (connector) {
+      case "prometheus":
+        storageEngine = null;
+        break;
+      default:
+        LOG.info(
+            "Unknown connector \"{}\". "
+                + "Please re-upload catalog configuration with a supported connector.",
+            catalog.get("connector"));
+        throw new IllegalStateException(
+            "Unknown connector. Connector doesn't exist in the list of supported.");
+    }
+    return storageEngine;
+  }
+
+  private void constructConnectors(ArrayNode catalogs) throws URISyntaxException {
```

Comment:
does CatalogServiceImpl multi thread safe?

Reply:
Yes. Do you see any problem? This method is private only called by SQLPlugin via loadConnectors method. The loadconnectors method is only available in plugin module and is invoked by SQLPlugin class. The other folders and classes only have instances of CatalogService. If you see CatalogService interface all the methods are read only.

Reply:
is it possible that get api return null catalog during catalog update. e.g. there are two connectors, "opensearch", "prometheus". 
1. user1 update prometheus.
2. user2 query opensearch, but catalog is empty now?

Reply:
Yeah will lock the objects while writing which is infrequent so wont have an impact.